File: frontend/web/data processing/features/receipt-parser/receipt_parser_core/receipt_recognizer.py
import os
from skimage.filters import threshold_local
from PIL import Image
import numpy as np
import cv2
from cv2.typing import MatLike
import matplotlib.pyplot as plt

BASE_PATH = os.getcwd()
INPUT_FOLDER = os.path.join(BASE_PATH, "data\\origin_img")
OUTPUT_FOLDER = os.path.join(BASE_PATH, "data\\img")


def prepare_folders():
    """
    :return: void
        Creates necessary folders
    """

    for folder in [INPUT_FOLDER, OUTPUT_FOLDER]:
        if not os.path.exists(folder):
            os.makedirs(folder)


def find_images(folder):
    """
    :param folder: str
        Path to folder to search
    :return: generator of str
        List of images in folder
    """

    for file in os.listdir(folder):
        full_path = os.path.join(folder, file)
        if os.path.isfile(full_path):
            try:
                _ = Image.open(full_path)  # if constructor succeeds
                yield file
            except:
                pass


def opencv_resize(image, ratio):
    width = int(image.shape[1] * ratio)
    height = int(image.shape[0] * ratio)
    dim = (width, height)
    return cv2.resize(image, dim, interpolation=cv2.INTER_AREA)


def plot_rgb(image):
    plt.figure(figsize=(16, 10))
    return plt.imshow(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))


def plot_gray(image):
    plt.figure(figsize=(16, 10))
    return plt.imshow(image, cmap="Greys_r")


def bw_scanner(image) -> MatLike:
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    sharpen_kernel = np.array([[-1, -1, -1], [-1, 9, -1], [-1, -1, -1]])
    sharpen = cv2.filter2D(gray, -1, sharpen_kernel)
    thresh = cv2.threshold(sharpen, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)[1]
    return cv2.bitwise_not(thresh)


# approximate the contour by a more primitive polygon shape
def approximate_contour(contour):
    epsilon = 0.01 * cv2.arcLength(contour, True)
    return cv2.approxPolyDP(contour, epsilon, True)


def get_receipt_contour(contours):
    for c in contours:
        approx = approximate_contour(c)
        if len(approx) == 4:
            return approx
        else:
            rect = cv2.minAreaRect(c)
            box = cv2.boxPoints(rect)
            box = np.intp(box)
            return approx


def contour_to_rect(contour, resize_ratio):
    x, y, w, h = cv2.boundingRect(contour)
    rect_points = np.array([[x, y], [x + w, y], [x + w, y + h], [x, y + h]])

    pts = rect_points.reshape(4, 2, order="F")
    rect = np.zeros((4, 2), dtype="float32")
    # top-left point has the smallest sum
    # bottom-right has the largest sum
    s = pts.sum(axis=1)
    rect[0] = pts[np.argmin(s)]
    rect[2] = pts[np.argmax(s)]
    # compute the difference between the points:
    # the top-right will have the minumum difference
    # the bottom-left will have the maximum difference
    diff = np.diff(pts, axis=1)
    rect[1] = pts[np.argmin(diff)]
    rect[3] = pts[np.argmax(diff)]
    return rect / resize_ratio


def wrap_perspective(img, rect) -> MatLike:
    # unpack rectangle points: top left, top right, bottom right, bottom left
    (tl, tr, br, bl) = rect
    # compute the width of the new image
    widthA = np.sqrt(((br[0] - bl[0]) ** 2) + ((br[1] - bl[1]) ** 2))
    widthB = np.sqrt(((tr[0] - tl[0]) ** 2) + ((tr[1] - tl[1]) ** 2))
    # compute the height of the new image
    heightA = np.sqrt(((tr[0] - br[0]) ** 2) + ((tr[1] - br[1]) ** 2))
    heightB = np.sqrt(((tl[0] - bl[0]) ** 2) + ((tl[1] - bl[1]) ** 2))
    # take the maximum of the width and height values to reach
    # our final dimensions
    maxWidth = max(int(widthA), int(widthB))
    maxHeight = max(int(heightA), int(heightB))
    # destination points which will be used to map the screen to a "scanned" view
    dst = np.array(
        [[0, 0], [maxWidth - 1, 0], [maxWidth - 1, maxHeight - 1], [0, maxHeight - 1]],
        dtype="float32",
    )
    # calculate the perspective transform matrix
    M = cv2.getPerspectiveTransform(rect, dst)
    # warp the perspective to grab the screen
    return cv2.warpPerspective(img, M, (maxWidth, maxHeight))


def get_basic_contour():
    pass


def get_complicated_contour():
    pass


def main():
    prepare_folders()
    images = list(find_images(INPUT_FOLDER))

    for image in images:
        input_path = os.path.join(INPUT_FOLDER, image)
        output_path = os.path.join(OUTPUT_FOLDER, image)
        image = cv2.imread(input_path)
        resize_ratio = 500 / image.shape[0]
        original = image.copy()
        image = opencv_resize(image, resize_ratio)
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        blurred = cv2.GaussianBlur(gray, (5, 5), 0)
        rectKernel = cv2.getStructuringElement(cv2.MORPH_RECT, (9, 9))
        dilated = cv2.dilate(blurred, rectKernel, iterations=1)
        closing = cv2.morphologyEx(dilated, cv2.MORPH_CLOSE, rectKernel)
        edged = cv2.Canny(closing, 50, 200, apertureSize=3)

        contours, hierarchy = cv2.findContours(
            edged, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE
        )
        largest_contours = sorted(contours, key=cv2.contourArea, reverse=True)[:10]
        receipt_contour = get_receipt_contour(largest_contours)
        # print(receipt_contour)
        try:
            scanned = wrap_perspective(
                original.copy(), contour_to_rect(receipt_contour, resize_ratio)
            )
        except AttributeError:
            print(input_path)
            continue

        result = bw_scanner(scanned)
        minval = np.percentile(result, 4.5)
        maxval = np.percentile(result, 95)
        pixvals = np.clip(result, minval, maxval)
        pixvals = ((pixvals - minval) / (maxval - minval)) * 255
        result = np.rot90(pixvals.astype(np.uint8), 1)
        cv2.imwrite(output_path, result)


if __name__ == "__main__":
    main()

File: lib/services/parser/receipt_recognizer.dart
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'package:frienance/utils/mat_extensions.dart';
import 'package:path/path.dart' as path;
import 'package:opencv_dart/opencv_dart.dart' as cv2;
import 'package:collection/collection.dart';

// Debug mode flag for pure Dart execution
const bool kDebugMode = bool.fromEnvironment('dart.vm.product') == false;

typedef Point2f = cv2.Point2f;

class ReceiptRecognizer {
  String basePath = 'D:\\Side Ultimatum\\Frienance\\lib\\cache';
  final String inputFolder = "1_source_img";
  final String outputFolder = "2_temp_img";

  static Future<ReceiptRecognizer> create() async {
    final instance = ReceiptRecognizer._();
    await instance._init();
    return instance;
  }

  ReceiptRecognizer._();

  Future<void> _init() async {
    // Use current working directory for pure Dart execution
    basePath = path.join(Directory.current.path, 'lib', 'cache');

    await prepareFolders();
  }

  Future<void> copyImagesToSourceDir(List<String> imagePaths) async {
    try {
      final sourceDir = Directory(path.join(basePath, inputFolder));
      await sourceDir.create(recursive: true);

      for (String imagePath in imagePaths) {
        final fileName = path.basename(imagePath);
        final destination = path.join(sourceDir.path, fileName);

        // Read directly from file system
        // Read directly from file system
        final bytes = await File(imagePath).readAsBytes();

        await File(destination).writeAsBytes(bytes);

        if (kDebugMode) {
          print('Copied $fileName to $destination');
        }
      }
    } catch (e) {
      if (kDebugMode) {
        print('Error copying images: $e');
      }
      rethrow;
    }
  }

  Future<void> prepareFolders() async {
    try {
      // Create base cache directory
      await Directory(basePath).create(recursive: true);

      // Create input and output directories
      final inputPath = path.join(basePath, inputFolder);
      final outputPath = path.join(basePath, outputFolder);

      await Future.wait([
        Directory(inputPath).create(recursive: true),
        Directory(outputPath).create(recursive: true),
      ]);

      if (kDebugMode) {
        print('Cache directory: $basePath');
        print('Input directory: $inputPath');
        print('Output directory: $outputPath');
      }
    } catch (e) {
      if (kDebugMode) {
        print('Error preparing folders: $e');
      }
      rethrow;
    }
  }

  Future<List<String>> findImages() async {
    try {
      final inputPath = path.join(basePath, inputFolder);

      if (!Directory(inputPath).existsSync()) {
        if (kDebugMode) {
          print('Input directory not found: $inputPath');
        }
        return [];
      }

      final images = Directory(inputPath)
          .listSync()
          .where((entity) =>
              entity is File &&
              ['.jpg', '.jpeg', '.png']
                  .contains(path.extension(entity.path).toLowerCase()))
          .map((entity) => entity.path)
          .toList();

      if (kDebugMode) {
        print('Found ${images.length} images in $inputPath');
      }

      return images;
    } catch (e) {
      if (kDebugMode) {
        print('Error finding images: $e');
      }
      return [];
    }
  }

  cv2.Mat opencvResize(cv2.Mat image, double ratio) {
    int width = (image.width * ratio).toInt();
    int height = (image.height * ratio).toInt();
    final dim = (width, height);
    return cv2.resize(image, dim, interpolation: cv2.INTER_AREA);
  }

  cv2.VecPoint orderPoints(cv2.VecPoint points) {
    // Order points as: top-left, top-right, bottom-right, bottom-left
    var pts = points.toList();
    
    // Sort by sum (x + y) - smallest is top-left, largest is bottom-right
    var sums = pts.map((p) => p.x + p.y).toList();
    var topLeft = pts[sums.indexOf(sums.reduce(min))];
    var bottomRight = pts[sums.indexOf(sums.reduce(max))];
    
    // Sort by difference (y - x) - smallest is top-right, largest is bottom-left
    var diffs = pts.map((p) => p.y - p.x).toList();
    var topRight = pts[diffs.indexOf(diffs.reduce(min))];
    var bottomLeft = pts[diffs.indexOf(diffs.reduce(max))];
    
    return cv2.VecPoint.fromList([topLeft, topRight, bottomRight, bottomLeft]);
  }

  cv2.Mat wrapPerspective(cv2.Mat img, cv2.VecPoint rect, double resizeRatio) {
    // Scale points back to original image size
    var scaledRect = cv2.VecPoint.fromList(
      rect.toList().map((p) => cv2.Point(
        (p.x / resizeRatio).round(),
        (p.y / resizeRatio).round()
      )).toList()
    );
    
    // Order the points correctly
    var orderedRect = orderPoints(scaledRect);
    
    // Unpack rectangle points
    var tl = orderedRect[0];
    var tr = orderedRect[1];
    var br = orderedRect[2];
    var bl = orderedRect[3];

    // Calculate width and height using Euclidean distance
    double widthA = sqrt(pow(br.x - bl.x, 2) + pow(br.y - bl.y, 2)); 
    double widthB = sqrt(pow(tr.x - tl.x, 2) + pow(tr.y - tl.y, 2));
    double heightA = sqrt(pow(tr.x - br.x, 2) + pow(tr.y - br.y, 2));
    double heightB = sqrt(pow(tl.x - bl.x, 2) + pow(tl.y - bl.y, 2));

    double maxWidth = max(widthA, widthB);
    double maxHeight = max(heightA, heightB);

    // Create destination points
    List<Point2f> dst = [
      Point2f(0, 0),
      Point2f(maxWidth, 0),
      Point2f(maxWidth, maxHeight),
      Point2f(0, maxHeight),
    ];

    // Perform perspective transformation
    final temp = cv2.getPerspectiveTransform(
        orderedRect,
        cv2.VecPoint.fromList(
            dst.map((p) => cv2.Point(p.x.toInt(), p.y.toInt())).toList()));
    return cv2.warpPerspective(img, temp, (maxWidth.toInt(), maxHeight.toInt()));
  }

  cv2.VecPoint approximateContour(cv2.VecPoint contour) {
    // 1. Get the convex hull to "straighten" the edges
    var hull = cv2.convexHull(contour); 
    // Convert mat to cv2.VecPoint 
    var hullVec = cv2.VecPoint.fromMat(hull);
    // 2. Then approximate from the hull, not the raw contour
    double epsilon = 0.01 * cv2.arcLength(hullVec, true); 
    return cv2.approxPolyDP(hullVec, epsilon, true);
  }

  cv2.VecPoint getReceiptContour(List<cv2.VecPoint> contours) {
    for (var contour in contours) {
      var approx = approximateContour(contour);
      
      if (approx.length == 4) {
        return approx;
      }
    }
    return cv2.VecPoint(); // Return empty VecPoint if none found
  }


  cv2.VecPoint contourToRect(cv2.VecPoint contour, double resizeRatio) {
    final boundingRect = cv2.boundingRect(contour);
    final x = boundingRect.x;
    final y = boundingRect.y;
    final w = boundingRect.width;
    final h = boundingRect.height;

    // Corner points
    List<List<double>> rectPoints = [
      [x.toDouble(), y.toDouble()],
      [x.toDouble() + w, y.toDouble()],
      [x.toDouble() + w, y.toDouble() + h],
      [x.toDouble(), y.toDouble() + h],
    ];

    // Sums for top-left (min) / bottom-right (max)
    List<double> s = rectPoints.map((p) => p[0] + p[1]).toList();
    int minSumIndex = s.indexOf(s.reduce(min));
    int maxSumIndex = s.indexOf(s.reduce(max));

    // Differences for top-right (min) / bottom-left (max)
    List<double> diff = rectPoints.map((p) => p[1] - p[0]).toList();
    int minDiffIndex = diff.indexOf(diff.reduce(min));
    int maxDiffIndex = diff.indexOf(diff.reduce(max));

    // Arrange points
    List<List<double>> rect = List.generate(4, (_) => List.filled(2, 0.0));
    rect[0] = rectPoints[minSumIndex];
    rect[2] = rectPoints[maxSumIndex];
    rect[1] = rectPoints[minDiffIndex];
    rect[3] = rectPoints[maxDiffIndex];

    // Scale by resizeRatio
    List<List<double>> scaled =
        rect.map((p) => [p[0] / resizeRatio, p[1] / resizeRatio]).toList();

    // Convert back to cv2.VecPoint
    return cv2.VecPoint.fromList(
      scaled.map((p) => cv2.Point(p[0].round(), p[1].round())).toList(),
    );
  }

  Future<String> saveToOutput(cv2.Mat image, String originalFileName,
      {String? suffix}) async {
    try {
      final fileName = suffix != null
          ? '${path.basenameWithoutExtension(originalFileName)}_$suffix${path.extension(originalFileName)}'
          : path.basename(originalFileName);

      final outputPath = path.join(basePath, outputFolder, fileName);

      await Directory(path.dirname(outputPath)).create(recursive: true);

      cv2.imwrite(outputPath, image);

      if (kDebugMode) {
        print('Saved image to: $outputPath');
      }

      return outputPath;
    } catch (e) {
      if (kDebugMode) {
        print('Error saving image: $e');
      }
      rethrow;
    }
  }

  double percentile(cv2.Mat data, double p) {
    if (data.isEmpty) {
      throw ArgumentError('Data cannot be empty');
    }
    if (p < 0 || p > 100) {
      throw ArgumentError('Percentile must be between 0 and 100');
    }

    // Flatten the cv2.Mat to a List<num>
    List<num> dataList = [];
    for (int i = 0; i < data.rows; i++) {
      for (int j = 0; j < data.cols; j++) {
        dataList.add(data.atNum(i, j));
      }
    }

    return dataList.sorted().percentile(p);
  }

cv2.Mat processImage(cv2.Mat result) {
  // Ensure we are working with Grayscale
  if (result.channels > 1) {
    result = cv2.cvtColor(result, cv2.COLOR_BGR2GRAY);
  }

  double minval = percentile(result, 4.5);
  double maxval = percentile(result, 95);

  // 1. Clip the values using OpenCV's thresholding/clipping
  // Anything below minval becomes minval; anything above maxval becomes maxval
  cv2.Mat clipped = result.clone();
  cv2.threshold(clipped, maxval, maxval, cv2.THRESH_TRUNC); // Cap at max
  cv2.threshold(clipped, minval, minval, cv2.THRESH_TOZERO_INV); // This is complex manually

  // 2. USE NORMALIZE - This replaces your second loop entirely and is more robust
  cv2.Mat normalized = cv2.Mat.empty();
  cv2.normalize(
    clipped, 
    normalized, 
    alpha: 0, 
    beta: 255, 
    normType: cv2.NORM_MINMAX, 
    dtype: cv2.MatType.CV_8U
  );

  return normalized;
}

  Future<void> saveProcessingStep(
      cv2.Mat image, String originalFileName, String step) async {
    await saveToOutput(image, originalFileName, suffix: 'step_$step');
  }

  void processReceipts(String fileName) async {
    var images = await findImages();
    for (var imagePath in images) {
      try {
        var fileName = path.basename(imagePath);
        // Read image
        cv2.Mat image = cv2.imread(imagePath);
        double resizeRatio = 1024 / image.shape[0];
        cv2.Mat original = image.clone();

        // Resize image
        image = opencvResize(image, resizeRatio);

        // Process image
        cv2.Mat gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY);
        // await saveProcessingStep(gray, fileName, '1_grayscale');

        // cv2.Mat blurred = cv2.gaussianBlur(gray, (5, 5), 0);
        cv2.Mat filtered = cv2.bilateralFilter(gray, 9, 75, 75);
        // await saveProcessingStep(filtered, fileName, '2_filtered');

        cv2.Mat rectKernel = cv2.getStructuringElement(cv2.MORPH_RECT, (9, 9));


        cv2.Mat dilated = cv2.dilate(filtered, rectKernel);
        // await saveProcessingStep(dilated, fileName, '3_dilated');
        cv2.Mat closing = cv2.morphologyEx(
          dilated,
          cv2.MORPH_CLOSE,
          rectKernel,
          anchor: cv2.Point(-1, -1),
          iterations: 1,
          borderType: cv2.BORDER_REPLICATE,
        );
        // await saveProcessingStep(closing, fileName, '4_closing');

        cv2.Mat edged = cv2.canny(closing, 50, 125);
        // await saveProcessingStep(edged, fileName, '5_edged');

        // Find contours
        var (contours, hierarchy) = cv2.findContours(
          edged,
          cv2.RETR_EXTERNAL,
          cv2.CHAIN_APPROX_SIMPLE,
        );

        var temp = (contours.toList().sorted((a, b) {
          double areaA = cv2.contourArea(a);
          double areaB = cv2.contourArea(b);
          return areaB.compareTo(areaA);
        }));
        cv2.VecVecPoint sortedContours = cv2.VecVecPoint.fromList(temp
            .map((e) =>
                e.map((e2) => cv2.Point(e2.x.toInt(), e2.y.toInt())).toList())
            .toList());

        // cv2.Mat contoursImage = original.clone();
        // cv2.drawContours(
        //   contoursImage,
        //   cv2.VecVecPoint.fromVecPoint(sortedContours.first),
        //   -1,
        //   cv2.Scalar.green,
        //   thickness: 3,
        // );
        // await saveProcessingStep(contoursImage, fileName, '6_largest_contours');
        // Sort contours by area

        var largestContours = sortedContours.take(10).toList(); // Take top 10 largest contours
        
        // // Draw approximated contours for visualization
        // cv2.Mat approxImage = original.clone();
        // for (var contour in largestContours) {
        //   var approx = approximateContour(contour);
        //   cv2.drawContours(
        //     approxImage,
        //     cv2.VecVecPoint.fromVecPoint(approx),
        //     -1,
        //     cv2.Scalar.blue,
        //     thickness: 3,
        //   );
        //   // Draw corner points in red
        //   for (var point in approx.toList()) {
        //     cv2.circle(
        //       approxImage,
        //       point,
        //       5,
        //       cv2.Scalar.red,
        //       thickness: -1,
        //     );
        //   }
        // }
        // await saveProcessingStep(approxImage, fileName, '7_approximated_contours');
        
        var receiptContour = getReceiptContour(largestContours);
        if (receiptContour.length != 4) {
          if (kDebugMode) {
            print('No 4-point receipt contour found in: $imagePath');
          }
          continue;
        }
        // var finalContour = original.clone();
        //   cv2.drawContours(
        //     finalContour,
        //     cv2.VecVecPoint.fromVecPoint(receiptContour),
        //     -1,
        //     cv2.Scalar.green,
        //     thickness: 3,
        //   );
        //   // Draw corner points in red
        //   for (var point in receiptContour.toList()) {
        //     cv2.circle(
        //       finalContour,
        //       point,
        //       5,
        //       cv2.Scalar.red,
        //       thickness: -1,
        //     );
        //   }
        
        // await saveProcessingStep(finalContour, fileName, '7_final_contour');
        var scanned = wrapPerspective(
            original, receiptContour, resizeRatio);
        // await saveProcessingStep(scanned, fileName, '7_scanned');
        var result = processImage(scanned);
        // await saveProcessingStep(result, fileName, '7_scanned');
        final savedPath = await saveToOutput(result, path.basename(imagePath),
            suffix: 'processed');
          print('Saved processed image to: $savedPath');
      } catch (e) {
          print('Error processing $imagePath: $e');
      }
    }
  }

}

Future<void> preworkImage() async {
  final recognizer = await ReceiptRecognizer.create();
  try {
    List<String> imageAssets = [];

    // Load images from assets/images directory (pure Dart approach)
    final assetsDir = Directory(path.join(Directory.current.path, 'assets', 'images'));
    if (assetsDir.existsSync()) {
      imageAssets = assetsDir
          .listSync(recursive: false)
          .whereType<File>()
          .where((f) => ['.jpg', '.jpeg', '.png']
              .contains(path.extension(f.path).toLowerCase()))
          .map((f) => f.path)
          .toList();
    }

    if (imageAssets.isEmpty) {
      throw StateError(
        'No images found to import. Place images in assets/images/ directory.',
      );
    }

    await recognizer.copyImagesToSourceDir(imageAssets);

    if (kDebugMode) {
      print('Imported ${imageAssets.length} images');
    }
  } catch (e) {
    if (kDebugMode) {
      print('Error importing assets: $e');
    }
    rethrow;
  }
  recognizer.processReceipts('');
}

void main() {
  preworkImage();
}
